### PR TITLE
support python3.5

### PR DESCRIPTION
**Pull request by jiapinai**
### WHY
I want to convert Pytorch model to support TensorRT in product envs while using the Pipeline.
And trying to build a docker image including both [TensorRT Inference Server](url) and Kale. Turns out, Kale requires **Python>3.6** while  TensorRT only supporting **Python2.7 and Python3.5** 

> [There are currently python wheels for python 2.7 and 3.5. Any chance of getting a wheel for 3.6?](https://devtalk.nvidia.com/default/topic/1035558/tensorrt/tensorrt-for-python-3-6-/)

As I'm not a Python expert, and the Kale code seems more friendly, the idea is to adjust Kale's code to support Python 3.5.  

### HOW
First, the way formatting string:

            `par_name = f"vol_{v['mount_point'].replace('/', '_').strip('_')}"`
Changed to

            `par_name = "vol_{}".format(v['mount_point'].replace('/', '_').strip('_'))`

Then, the way using file Path:

        `output_path = os.path.join(os.path.dirname(self.source_path),`
Changed to

        `output_path = os.path.join(os.path.dirname(self.source_path.__str__()),`

###Tests

I have tested the code by running the examples [pytorch-classification](https://github.com/kubeflow-kale/examples) and it works like a charm

### CODE REVIEW

**Comment by StefanoFioravanzo** [comment]
@jiapinai Thank you for contributing to the project!

Can you please rebase on top of branch `feature-refactoring-and-tests`? We are refactoring the codebase and we will merge that branch to `master` very soon. You can change the base branch of the PR by clicking on `Edit` on top of this page :)

**Comment by jiapinai** [comment]
sure, rebased, and trying to resolve those conflicts

File: kale/codegen/generate_code.py [hunk 1]
```diff
@@ -32,15 +32,15 @@ def get_volume_parameters(volumes):
             continue
 
         if v['type'] == 'pvc':
-            par_name = f"vol_{v['mount_point'].replace('/', '_').strip('_')}"
+            par_name = "vol_{}".format(v['mount_point'].replace('/', '_').strip('_'))
```
Comment (StefanoFioravanzo) [inline kale/codegen/generate_code.py]: let's not break the 80 chars limit:

```suggestion
            mount_point = v['mount_point'].replace('/', '_').strip('_')
            par_name = "vol_{}".format(mount_point)
```

File: kale/codegen/generate_code.py [hunk 2]
```diff
@@ -74,7 +74,7 @@ def get_marshal_data(wd, volumes, nb_path):
         # if we found any, then set marshal directory inside working directory
         if len(vols) > 0:
             marshal_volume = False
-            marshal_dir = f".{os.path.basename(nb_path)}.kale.marshal.dir"
+            marshal_dir = ".{}.kale.marshal.dir".format(os.path.basename(nb_path))
```
Comment (StefanoFioravanzo) [inline kale/codegen/generate_code.py]:
```suggestion
            basename = os.path.basename(nb_path)
            marshal_dir = ".{}.kale.marshal.dir".format(basename)
```

File: kale/codegen/generate_code.py [hunk 3]
```diff
@@ -98,12 +98,12 @@ def get_args(pipeline_parameters):
     """
     pipeline_args_names = ', '.join(list(pipeline_parameters.keys()))
     # wrap in quotes every parameter - required by kfp
-    pipeline_args = ', '.join([f"{arg}='{pipeline_parameters[arg][1]}'"
+    pipeline_args = ', '.join(["{}='{}'".format(arg, pipeline_parameters[arg][1])
```
Comment (StefanoFioravanzo) [inline kale/codegen/generate_code.py]:
```suggestion
    pipeline_args = ', '.join(["{}='{}'".format(arg, 
                                                pipeline_parameters[arg][1])
```

File: kale/codegen/generate_code.py [hunk 4]
```diff
@@ -98,12 +98,12 @@ def get_args(pipeline_parameters):
...
                                for arg in pipeline_parameters])
     # Arguments are the pipeline arguments. Since we don't know precisely in
     # what pipeline steps they are needed, we just pass them to every one.
     # We assume there variables were not re-assigned throughout the notebook
-    function_args = ', '.join([f"{arg}: {pipeline_parameters[arg][0]}"
+    function_args = ', '.join(["{}: {}".format(arg, pipeline_parameters[arg][0])
```
Comment (StefanoFioravanzo) [inline kale/codegen/generate_code.py]:
```suggestion
    function_args = ', '.join(["{}: {}".format(arg, 
                                               pipeline_parameters[arg][0])
```

File: kale/rpc/nb.py [hunk 5]
```diff
@@ -72,7 +72,7 @@ def compile_notebook(request, source_notebook_path,
 def _get_kale_marshal_dir(source_notebook_path):
     nb_file_name = os.path.basename(source_notebook_path)
     nb_dir_name = os.path.dirname(source_notebook_path)
-    kale_marshal_dir_name = f".{nb_file_name}{KALE_MARSHAL_DIR_POSTFIX}"
+    kale_marshal_dir_name = ".{}{}".format(nb_file_name, KALE_MARSHAL_DIR_POSTFIX)
```
Comment (StefanoFioravanzo) [inline kale/rpc/nb.py]:
```suggestion
    kale_marshal_dir_name = ".{}{}".format(nb_file_name, 
                                           KALE_MARSHAL_DIR_POSTFIX)
```

File: kale/rpc/rok.py [hunk 6]
```diff
@@ -77,7 +77,7 @@ def _get_cloned_volume(volume, obj_name, members):
                                       'value': member['rok_url']}]
             return volume
 
-    msg = f"Volume '{volume['name']}' not found in group '{obj_name}'"
+    msg = "Volume '{}' not found in group '{}'".format(volume['name'], obj_name)
```
Comment (StefanoFioravanzo) [inline kale/rpc/rok.py]:
```suggestion
    msg = "Volume '{}' not found in group '{}'".format(volume['name'], 
                                                       obj_name)
```

File: kale/static_analysis/inspector.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
+import ast
```
Comment (StefanoFioravanzo) [inline kale/static_analysis/inspector.py]: This file should not be included in the PR. It was delete in a previous PR. You might have included it during the merge by mistake

File: kale/utils/kfp_utils.py [hunk 8]
```diff
@@ -74,7 +74,7 @@ def upload_pipeline(pipeline_package_path, pipeline_name, overwrite=False, host=
         # The exception is a general 500 error.
         # The only way to check that it refers to the pipeline already existing
         # is by matching the error message
-        if overwrite and f'The name {pipeline_name} already exist' in str(e):
+        if overwrite and 'The name {} already exist'.format(pipeline_name) in str(e):
```
Comment (StefanoFioravanzo) [inline kale/utils/kfp_utils.py]:
```suggestion
        exc_msg = 'The name {} already exist'.format(pipeline_name)
        if overwrite and exc_msg in str(e):
```

File: kale/core.py [hunk 9]
```diff
@@ -143,7 +143,7 @@ def validate_metadata(self):
                 if 'name' not in v:
                     raise ValueError("Provide a valid name for every volume")
                 if not re.match(k8s_valid_name_regex, v['name']):
-                    raise ValueError(f"PV/PVC resource name {k8s_name_msg}")
+                    raise ValueError("PV/PVC resource name {}".format(k8s_name_msg))
```
Comment (StefanoFioravanzo) [inline kale/core.py]:
```suggestion
                    raise ValueError("PV/PVC resource name {}"
                                     .format(k8s_name_msg))
```

**Comment by StefanoFioravanzo** [comment]
I think it would also be nice to remove the usage of `pathlib` from `core.py`, so we can avoid having lots of `__str__()`.

**Comment by jiapinai** [comment]
> I think it would also be nice to remove the usage of `pathlib` from `core.py`, so we can avoid having lots of `__str__()`.


The changes are: 
1.  the file `inspector.py` is removed
2. the 'pathlib' in `core.py` is removed
3. most of the lines are limited to 80 chars now

File: kale/core.py [hunk 10]
```diff
@@ -243,7 +245,7 @@ def detect_environment(self):
         """
         # used to set container step working dir same as current environment
         self.pipeline_metadata['abs_working_dir'] = os.path.dirname(
-            os.path.abspath(self.source_path.__str__()))
+            os.path.abspath(self.source_path)
```
Comment (StefanoFioravanzo) [inline kale/core.py]:
```suggestion
            os.path.abspath(self.source_path))
```

**Comment by jiapinai** [comment]
my mistake, changed

**Comment by StefanoFioravanzo** [comment]
Ok thanks, LGTM :)
